### ISSUE TITLE
No error when generic cannot be inferred

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#11701]
v2.0.3

See actual/expected in code comments:

``` ts
    const foo = <T>(t1: T, t2: T) => {}
    type Bar = { baz: 'bob' }
    const bar: Bar = { baz: 'bob' }
    foo(bar, bar) // actual and expected: no error
    foo(bar, {}) // actual: no error, expected: error. generic is silently inferred as {} (unknown)
    foo(bar, { baz: 1 }) // actual and expected: error
```

Is there a way to disable the silent failure here? If my generic cannot be inferred, I want a compile error so I am forced to provide one.

**Comment by alitaheri** [comment]
A flag to enable this strictness would be very nice. although, maybe it can be added to `noImplicitAny`. But it might break a lot of code. How about `noGenericUnknown` O.o

**Comment by RyanCavanaugh** [comment]
It is inferred. What you're proposing would make a lot of perfectly valid code an error -- it's quite common to have code like this
```ts
let a: Animal = ...;
let d: Dog = ...;
foo(a, d);
```
Just as `Animal` is a valid supertype of `Animal` and `Dog`, `{}` is a valid supertype of `{}` and `{ baz: string }`.

**Comment by OliverJAsh** [comment]
I see, it's not inferring `{}` because it doesn't know what to infer, but rather because it is the common type?

In that case, I don't understand why this doesn't work:

```
    const foo2 = <T1, T2 extends T1>(t1: T1, t2: T2) => {}
    foo(bar, bar) // actual and expected: no error
    foo(bar, {}) // actual: no error, expected: error.
```

I basically want to say the second generic/parameter should always be a subtype of the first generic/parameter.

**Comment by HerringtonDarkholme** [comment]
You can use currying.
```typescript
const foo = <T>(t1: T) => (t2: T) => {}
```

**Comment by RyanCavanaugh** [comment]
> I basically want to say the second generic/parameter should always be a subtype of the first generic/parameter.

I don't think there's a way to get that behavior. The second parameter will always be an inference site and there's not really a way to turn that off. We've had some suggestions before for "non-inferential" generic use sites but the scenarios are pretty thin.
